pkg/inject/inject.go

```diff
@@ -0,0 +1,571 @@
+	//	   containers to be self contained.
+	//  4. We skip recording telemetry for intra-pod traffic within the control plane.
+
+	switch conf.meta.Kind {
+	case "Deployment":
+		var deployment v1beta1.Deployment
+		if err := yaml.Unmarshal(conf.bytes, &deployment); err != nil {
+			return err
+		}
+
+		if deployment.Name == ControlPlanePodName && deployment.Namespace == globalConfig.GetLinkerdNamespace() {
+			conf.dnsNameOverride = LocalhostDNSNameOverride
+		}
+
+		conf.obj = &deployment
+		conf.k8sLabels[k8s.ProxyDeploymentLabel] = deployment.Name
+		conf.podSpec = &deployment.Spec.Template.Spec
+		conf.objectMeta = &deployment.Spec.Template.ObjectMeta
+
+	case "ReplicationController":
+		var rc v1.ReplicationController
+		if err := yaml.Unmarshal(conf.bytes, &rc); err != nil {
+			return err
+		}
+
+		conf.obj = &rc
+		conf.k8sLabels[k8s.ProxyReplicationControllerLabel] = rc.Name
+		conf.podSpec = &rc.Spec.Template.Spec
+		conf.objectMeta = &rc.Spec.Template.ObjectMeta
+
+	case "ReplicaSet":
+		var rs v1beta1.ReplicaSet
+		if err := yaml.Unmarshal(conf.bytes, &rs); err != nil {
+			return err
+		}
+
+		conf.obj = &rs
+		conf.k8sLabels[k8s.ProxyReplicaSetLabel] = rs.Name
+		conf.podSpec = &rs.Spec.Template.Spec
+		conf.objectMeta = &rs.Spec.Template.ObjectMeta
+
+	case "Job":
+		var job batchV1.Job
+		if err := yaml.Unmarshal(conf.bytes, &job); err != nil {
+			return err
+		}
+
+		conf.obj = &job
+		conf.k8sLabels[k8s.ProxyJobLabel] = job.Name
+		conf.podSpec = &job.Spec.Template.Spec
+		conf.objectMeta = &job.Spec.Template.ObjectMeta
+
+	case "DaemonSet":
+		var ds v1beta1.DaemonSet
+		if err := yaml.Unmarshal(conf.bytes, &ds); err != nil {
+			return err
+		}
+
+		conf.obj = &ds
+		conf.k8sLabels[k8s.ProxyDaemonSetLabel] = ds.Name
+		conf.podSpec = &ds.Spec.Template.Spec
+		conf.objectMeta = &ds.Spec.Template.ObjectMeta
+
+	case "StatefulSet":
+		var statefulset appsV1.StatefulSet
+		if err := yaml.Unmarshal(conf.bytes, &statefulset); err != nil {
+			return err
+		}
+
+		conf.obj = &statefulset
+		conf.k8sLabels[k8s.ProxyStatefulSetLabel] = statefulset.Name
+		conf.podSpec = &statefulset.Spec.Template.Spec
+		conf.objectMeta = &statefulset.Spec.Template.ObjectMeta
+
+	case "Pod":
+		var pod v1.Pod
+		if err := yaml.Unmarshal(conf.bytes, &pod); err != nil {
+			return err
+		}
+
+		conf.obj = &pod
+		conf.podSpec = &pod.Spec
+		conf.objectMeta = &pod.ObjectMeta
+	}
+
+	return nil
+}
+
+// Given a PodSpec, update the PodSpec in place with the sidecar
+// and init-container injected. If the pod is unsuitable for having them
+// injected, return false.
+func injectPodSpec(t *v1.PodSpec, patch *Patch, identity k8s.TLSIdentity, controlPlaneDNSNameOverride string, globalConfig *pb.GlobalConfig, proxyConfig *pb.ProxyConfig, report *Report) bool {
+	report.HostNetwork = t.HostNetwork
+	report.Sidecar = healthcheck.HasExistingSidecars(t)
+	report.Udp = checkUDPPorts(t)
```

Comment:
TIOLI: I think the report-related code don't need to be here. Similar observation for `injectObjectMeta()` and `uninjectPodSpec()`. If all these report code are moved to the parent `Transform()` and `Uninject()` methods, we can potentially create inject-specific and uninject-specific reports in the `Report` struct, and returned them to the CLI. Hence, trimming down the current `generateReport()` functions.

Reply:
Agreed. It appears it wouldn't be to hard to extract that out :+1:

Reply:
This was partially addressed in #2379 and back-ported here